Separate AppInitStartClients from AppInitMain

hebasto opened the pull request:
This change allows to make the GUI startup process more granular, and, eventually, to move the `appInitStartClients` call from
`BitcoinCore::initialize` into `LoadWalletsActivity` (see https://github.com/bitcoin-core/gui/pull/342) and fix https://github.com/bitcoin-core/gui/issues/247.

~No behavior change.~

hebasto committed:
refactor: Separate AppInitStartClients from AppInitMain

This change allows to make the GUI startup process more granular, and,
eventually, to move the appInitStartClients call from
BitcoinCore::initialize into another thread with better user feedback.

No behavior change.

MarcoFalke commented on src/qt/bitcoin.cpp:
```diff
@@ -174,6 +174,7 @@ void BitcoinCore::initialize()
         qDebug() << __func__ << ": Running initialization in thread";
         interfaces::BlockAndHeaderTipInfo tip_info;
         bool rv = m_node.appInitMain(&tip_info);
+        m_node.appInitStartClients();
```
I don't think this is supposed to run when the previous call failed

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/22231#issuecomment-860202506).

hebasto commented:
Updated e734449b8025b7e5a7c184672a910737b9420706 -> 3268fcc6d65d2f35313a314a6560839715fe602c ([pr22231.01](https://github.com/hebasto/bitcoin/commits/pr22231.01) -> [pr22231.02](https://github.com/hebasto/bitcoin/commits/pr22231.02), [diff](https://github.com/hebasto/bitcoin/compare/pr22231.01..pr22231.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/22231#discussion_r650472800)

hebasto commented:
Rebased 3268fcc6d65d2f35313a314a6560839715fe602c -> 32aad05fe88f12bf7ff5ce3acb62e6f06d944bac ([pr22231.02](https://github.com/hebasto/bitcoin/commits/pr22231.02) -> [pr22231.03](https://github.com/hebasto/bitcoin/commits/pr22231.03)) due to the conflict with https://github.com/bitcoin-core/gui/pull/381.

ryanofsky reviewed (approved):
This seems ok. I tend to think an alternate approach where bitcoin node and wallet starts up by itself and sends more frequent notifications to the GUI would be better than having GUI code drive bitcoin node and wallet loading from the outside like this.

Code review ACK 3268fcc6d65d2f35313a314a6560839715fe602c in any case

MarcoFalke commented:
How is this a refactor when the behaviour changes? Previously `StartupNotify` would be called after the wallets are started, now it is called after.

ryanofsky commented:
> How is this a refactor when the behaviour changes? Previously `StartupNotify` would be called after the wallets are started, now it is called before.

I'm not sure this is a problem, since I think it's just moving wallet postInit calls, not wallet loading calls. So wallets should still be loaded and RPCs should function. But this would be good to mention in the PR description. And if there actually are user-visible changes in behavior, it would probably be good to add release notes and update `-startupnotify` documentation, or to update the PR to avoid them.

> This seems ok. I tend to think an alternate approach where bitcoin node and wallet starts up by itself and sends more frequent notifications to the GUI would be better than having GUI code drive bitcoin node and wallet loading from the outside like this.

Note that 71d524015b57610e50f62da57497dfa020855445 from #10102 makes a similar change as this PR, but takes the alternate approach of adding a new init notification instead of adding a new init method, so common init code is driving itself and notifying the GUI, instead of the GUI driving the init.

71d524015b57610e50f62da57497dfa020855445 and this PR are both complementary and both have a similar motivation of making GUI startup more responsive. For comparison, 71d524015b57610e50f62da57497dfa020855445 moves WalletClient construction (wallet process spawning) from the GUI thread to the init executor thread, while this PR moves WalletClient::start calls from earlier in init executor thread to later in the same thread.

MarcoFalke commented:
It is absolutely a behaviour change if `-startupnotify` previously waited for the wallet txs mempool state to be updated, and now it doesn't.

ryanofsky commented:
> It is absolutely a behaviour change if `-startupnotify` previously waited for the wallet txs mempool state to be updated, and now it doesn't.

Makes sense and I also wonder if this doesn't really have an impact on bitcoin-core/gui#247, contrary to the description:

> This change allows to make the GUI startup process more granular, and, eventually, to move the `appInitStartClients` call from
> `BitcoinCore::initialize` into `LoadWalletsActivity` (see [bitcoin-core/gui#342](https://github.com/bitcoin-core/gui/pull/342)) and fix the [bitcoin-core/gui#247](https://github.com/bitcoin-core/gui/issues/247).
> 
> No behavior change.

This PR only affects wallet postinit (client->start), not wallet loading (client->load), or GUI WalletModel loading.

hebasto commented:
@ryanofsky 
> > It is absolutely a behaviour change if `-startupnotify` previously waited for the wallet txs mempool state to be updated, and now it doesn't.
> 
> Makes sense and I also wonder if this doesn't really have an impact on [bitcoin-core/gui#247](https://github.com/bitcoin-core/gui/issues/247), contrary to the description:
> 
> > This change allows to make the GUI startup process more granular, and, eventually, to move the `appInitStartClients` call from
> > `BitcoinCore::initialize` into `LoadWalletsActivity` (see [bitcoin-core/gui#342](https://github.com/bitcoin-core/gui/pull/342)) and fix the [bitcoin-core/gui#247](https://github.com/bitcoin-core/gui/issues/247).
> > No behavior change.
> 
> This PR only affects wallet postinit (client->start), not wallet loading (client->load), or GUI WalletModel loading.

You're [right](https://github.com/bitcoin-core/gui/pull/342#issuecomment-917628717). This PR does not have an impact on https://github.com/bitcoin-core/gui/issues/247.

Closing.
